libs/threading_base/src/register_thread.cpp

```diff
@@ -39,6 +56,22 @@ namespace hpx { namespace threads { namespace detail {
         }
         else
         {
+
+// The following implementation has been divided for Linux and Mac OSX
+#if defined(HPX_HAVE_DYNAMIC_HPX_MAIN) && \
+    (defined(__linux) || defined(__linux__) || defined(linux) || \
+    defined(__APPLE__))
+
+            // hpx_main.hpp is included but not linked to libhpx_wrap
+            if (!hpx_start::is_linked && hpx_start::include_libhpx_wrap)
+                HPX_THROW_EXCEPTION(invalid_status,
+                    "hpx::threads::detail::get_self_or_default_pool",
+                    "Attempting to use hpx_main.hpp functionality without "
+                    "linking to libhpx_wrap. Did you link to HPX::wrap_main "
+                    "in your CMakeLists.txt?");
```

Comment:
Referring to `CMakeLists.txt` might be misleading as -- while this is the preferred way of linking HPX applications -- users might chose to directly link without using CMake.

Reply:
I've added a verbose exception message to cater to users using `CMakeLists.txt` and `Makefile`. `pkg-config` takes care of the linking part so we're good there.

Reply:
What about users that take care of the linking explicitly (see #4880) and get it wrong anyways?

Reply:
@hkaiser don't think its a good idea to link explicitly (unless no other way). Should I just add a line that check HPX docs for instructions if linking manually?

Reply:
That might be sufficient, indeed. Thanks!

Reply:
done!